interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package requestprompts
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces/prompting"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify/listener"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// Prompt contains information about a request for which a user should be
+// prompted.
+type Prompt struct {
+	ID           string             `json:"id"`
+	Timestamp    time.Time          `json:"timestamp"`
+	Snap         string             `json:"snap"`
+	Interface    string             `json:"interface"`
+	Constraints  *promptConstraints `json:"constraints"`
+	listenerReqs []*listener.Request
+}
+
+// promptConstraints are like prompting.Constraints, but have a "path" field
+// instead of a "path-pattern", and include the available permissions for the
+// interface corresponding to the prompt.
+type promptConstraints struct {
+	Path                 string   `json:"path"`
+	Permissions          []string `json:"permissions"`
```

Comment:
are those requested permissions? if so could the field be named eg `RequestedPermissions`

Reply:
They are the requested permissions. Iirc, these started out as `"requested-permissions"`, but we moved to `"permissions"` instead for some reason. I'd be happy to go back to `"requested-permissions"` if @pedronis agrees.

Reply:
you could call the field RequestedPermission why keeping the agreed external JSON name just permissions, otoh maybe the fact that the overall thing is Constrains is clear enough?

Reply:
Yes, the idea was to make `PromptConstraints` as similar as possible to `Constraints`, with the exception that we must use `path` instead of `path-pattern`. I think using the `Permissions` field throughout for both types is clearer, especially since the `Permissions` might get pruned over time as new rules satisfy some of the permissions, so `Permissions` is no longer matches the originally requested permissions. Also, `AvailablePermissions` is never used internally, it's just populated once as information for the client.

Furthermore, I'll soon be working on the ability to send back a bitmask of permissions corresponding to those provided in the user's reply, rather than just exactly what was asked for, and I think we'll need to record the original list of abstract permissions (not exported or included in the json) so that prompts which are partially satisfied by several new rules still send back a reply with all the requested permissions. So `RequestedPermissions` would be a bit ambiguous here whether it's referring to the originally requested permissions or the current remaining permissions.